File: packages/react-kivex/scripts/watch.ts
import path from 'path';
import { generateIcons } from './build';

const ICONS_DIRECTORY = path.join(path.resolve(__dirname, '..'), '../../icons');

async function watch() {
  const chokidar = await import('chokidar');

  // Initial run
  await generateIcons();

  console.log(`Watching ${ICONS_DIRECTORY} for changes...`);
  console.log('   Press Ctrl+C to stop.');

  let timeout: NodeJS.Timeout | null = null;
  let isRunning = false;

  const watcher = chokidar.watch(ICONS_DIRECTORY, {
    persistent: true,
    ignoreInitial: true,
    depth: 10,
  });

  watcher.on('all', (event, filePath) => {
    if (timeout) clearTimeout(timeout);
    timeout = setTimeout(() => {
      if (isRunning) return;
      isRunning = true;
      console.log(`\n🔄 Change detected (${event}), regenerating...`);
      generateIcons()
        .catch((err) => console.error('Error during regeneration:', err))
        .finally(() => {
          isRunning = false;
          timeout = null;
        });
    }, 200);
  });

  process.on('SIGINT', () => {
    console.log('\n👋 Stopping watch...');
    watcher.close();
    process.exit(0);
  });
}

if (require.main === module) {
  watch().catch((err) => {
    console.error('Error:', err);
    process.exit(1);
  });
}

File: packages/kivex_flutter/generator/watch.dart
import 'dart:async';
import 'dart:io';
import 'package:path/path.dart' as p;

void main(List<String> args) async {
  final force = args.contains('--force');

  final scriptDir = p.dirname(p.fromUri(Platform.script));
  final rootDir = p.dirname(scriptDir);
  final iconsDir = p.join(rootDir, '..', '..', 'icons');

  // Initial run
  await _runGenerator(force);
  print('👀 Watching $iconsDir for changes...');
  print('   Press Ctrl+C to stop.');

  Timer? debounce;
  bool isRunning = false;

  final watcher = Directory(iconsDir).watch(recursive: true);

  await for (final event in watcher) {
    // Only respond to file events (modify, create, delete)
    if (event.type != FileSystemEvent.modify &&
        event.type != FileSystemEvent.create &&
        event.type != FileSystemEvent.delete) {
      continue;
    }

    // Check if the affected path is an SVG file
    final eventPath = event.path;
    if (!eventPath.endsWith('.svg')) continue;

    debounce?.cancel();
    debounce = Timer(const Duration(milliseconds: 200), () async {
      if (isRunning) return;
      isRunning = true;
      print('\n🔄 Change detected (${event.type}), regenerating...');
      try {
        await _runGenerator(force);
      } catch (e) {
        print('❌ Error during regeneration: $e');
      } finally {
        isRunning = false;
        debounce = null;
      }
    });
  }
}

Future<void> _runGenerator(bool force) async {
  final args = ['run', 'generator/build.dart'];
  if (force) args.add('--force');

  final result = await Process.run('dart', args);
  if (result.exitCode != 0) {
    stderr.write(result.stderr);
    throw Exception('Generator failed with exit code ${result.exitCode}');
  } else {
    stdout.write(result.stdout);
  }
}
